bundles/org.openhab.binding.boschspexor/src/main/java/org/openhab/binding/boschspexor/internal/BoschSpexorThingHandler.java

```diff
@@ -0,0 +1,279 @@
+                updateState(getChannelID(GROUP_ID_STATUS, CHANNEL_ID_ENERGY_MODE),
+                        new StringType(energy.getEnergyMode().name()));
+                updateState(getChannelID(GROUP_ID_STATUS, CHANNEL_ID_POWERED), OnOffType.from(energy.isPowered()));
+                updateState(getChannelID(GROUP_ID_STATUS, CHANNEL_ID_SOC),
+                        new PercentType(energy.getStateOfCharge().getValue()));
+
+                // FIRMWARE INFO
+                Firmware firmware = spexor.getStatus().getFirmware();
+                updateState(getChannelID(GROUP_ID_STATUS, CHANNEL_ID_AVAILABLE_VERSION),
+                        new StringType(firmware.getAvailableVersion()));
+                updateState(getChannelID(GROUP_ID_STATUS, CHANNEL_ID_INSTALLED_VERSION),
+                        new StringType(firmware.getCurrentVersion()));
+                updateState(getChannelID(GROUP_ID_STATUS, CHANNEL_ID_FIRMWARE_STATE),
+                        new StringType(firmware.getState().name()));
+
+                // PROFILE
+                Profile profile = spexor.getProfile();
+                updateState(getChannelID(GROUP_ID_PROFILE, CHANNEL_ID_PROFILE_NAME), new StringType(profile.getName()));
+                updateState(getChannelID(GROUP_ID_PROFILE, CHANNEL_ID_PROFILE_TYPE),
+                        new StringType(profile.getProfileType().name()));
+
+                boolean thingStructureChanged = false;
+                ThingBuilder thingBuilder = editThing();
+
+                // OBSERVATION
+                for (ObservationStatus observationStatus : spexor.getStatus().getObservation()) {
+                    @SuppressWarnings("null")
+                    String observationType = observationStatus.getObservationType();
+                    Channel channel = getThing().getChannel(getChannelID(GROUP_ID_OBSERVATIONS, observationType));
+                    if (channel == null) {
+                        channel = createObservationChannel(observationType);
+                        thingBuilder.withoutChannel(channel.getUID()).withChannel(channel);
+                        thingStructureChanged = true;
+                    }
+                    updateState(channel.getUID(), new StringType(observationStatus.getSensorMode().name()));
+                }
+
+                // SENSORS
+                Map<String, SensorValue<?>> values = apiService.getSensorValues(spexor.getId(), spexor.getSensors());
+                for (String sensor : spexor.getSensors()) {
+                    String sensorType = sensor;
+                    Channel channel = getThing().getChannel(getChannelID(GROUP_ID_SENSORS, sensorType));
+                    if (channel == null) {
+                        channel = createSensorChannel(sensor, sensorType);
+                        if (channel != null) {
+                            thingBuilder.withoutChannel(channel.getUID()).withChannel(channel);
```

Comment:
Thanks. Then I do have a question regarding this. The spexor provides the possibility to add additional features after it got ready by the user. The user could by additional features -> new channels. The thing itself then has to be deleted by the user and again discovered?

Reply:
I would probably implement this via the discovery service. The discovery service has a reference to the thing via the bridge. So it can pass the found device and channels to the things connected to the bridge. It will then be a separate step from the update of the state in the thing. In that method add/remove/update the channels. That way the discovery doesn't need to recreate the thing.

Reply:
@Hilbrand thanks for your fast answer!!!
I checked it now multiple times and tried some options but my conclusion is ended up to the same:
It won't work with the idea of a spexor. The spexor is not a fully defined device and it can be configured / extended by a user or the product development team. That means new features would appear and every integrator of the spexor would need to implement another version of the connector.  

The bundle is written as generic as possible to have stable base and doesn't need to be adjusted quite often if the spexor provides new features, sensor etc.  Deleting a thing by a customer to get newer features seems for me also a draw back in user experience and usage of the openHAB.

I also verified a few other bindings how they have solved it and inside the yamahareceiver addon there is the similar (but more complexe) behavior defined with the same mechanism.

If I got it correct you mean channels have to be known in the initialization phase. If this is the case then there are two options I could now apply: 
a) user has to delete for every time the thing in openHAB and add this again -> not that nice user behavior
b) The PR code wouldn't get any approval with this because internal handling of openHAB would limit it.  A complete refactoring here would be necessary in this bundle to satisfy the fixed channel behavior.

Do you know an alternative solution where I can have a look?-

Reply:
> If I got it correct you mean channels have to be known in the initialization phase.

No it can be added any time. However from a user point of view new channels need to be configured. Therefor when the channels are changed it requires user interaction anyway to make use of them. From that point of view it makes more sense to have the update of channels be part of the discovery process (This can be either the background process or manually triggered by the user. The background discovery can be disabled by the user and thus if the user decides no need to check for new channels it can be disabled).

My suggestion is to add a separate public method to the `BoschSpexorThingHandler`. This method performs the change  to the channels. The part under the `OBSERVATION` comment. Since discovery has a link to the bridge it can call a method to the bridge to `discoverChannels` (or some name for the method) during a discovery run. Then the bridge can iterate over all things and also call `discoverChannels` on each thing. Then that method in the thing performs the code under the comment  `OBSERVATION` including the optional update of the thing.

For the update you need to have the `SpexorInfo` object. What you can do is use `ExpiringCache`. That is a cache object that has a method argument that takes care of the actual call to the service. Below I've copied in how that might look like:

```
  private final ExpiringCache<SpexorInfo> cache;

  --- in constructor
   cache = new ExpiringCache<>(3, this::getSpexorInfo);
  -- 

  private SpexorInfo getSpexorInfo() {
        SpexorAPIService apiService = getSpexorAPIService();
        if (apiService == null) {
            logger.warn("spexor API service is not available and device won't be updated");
        } else if (getSpexorID().isEmpty()) {
            logger.warn("thing is not available and device can't be updated");
        } else {
            SpexorInfo spexor = apiService.getSpexor(getSpexorID().get());
            if (spexor == null) {
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                        "Could not determine further information.");
           }
          return spexor;
      }
      return null;
  }
```
Then in both the channel update and discover method you call `cache.getValue()` and than it will only actual call the service when the data is expired as passed to the first parameter of `ExpiringCache` constructor. I've used `3` here. Because that way if the `getValue()` is called within 3 seconds it won't make an extra call. This is especially relevant because the `REFRESH` command is likely called on all linked channels. This means it can triggered several times after each other and in such case you don't want to trigger a call to the service again. But you might see what time is best for your case.

Reply:
@Hilbrand Okay I now changed the behavior as proposed. I exchanged your suggestion with the expiringcache a bit and bound it more flexible be the user itself. Thanks a lot hopefully its not fitting to the expectations. Thanks!